improve RichText v-align and support img align feature

RazgrizHsu opened the pull request:
小修正及部份改进，
代码在我的项目已经运行一段时间，没发现什么问题，
各位大神帮忙检视一下这样是否可行，谢谢

### Changes:
- 修正outline的垂直对齐误差
   fix v-align for`<outline width=2>text</outline>`

- 让`<img height=999>`设定值可以设定超过行高
   add`<img>`ability to set size over lineHeight

- 新增`<img align=top>`设定垂直对齐属性，可为top|center|bottom，未设定及预设值为bottom
   add`<img>`align tag like align=top|center|bottom, if not setting use default:bottom

### 未修改时的效果
![image](https://user-images.githubusercontent.com/3889141/71641626-5fcd3c80-2cda-11ea-9c98-7e16dffe2dd3.png)

### 修改后的效果
![image](https://user-images.githubusercontent.com/3889141/71641631-6a87d180-2cda-11ea-9e8b-997602cf0db8.png)

RazgrizHsu committed:
improve RichText for correct v-align and support img align feature

- fix v-align for <outline width=2>text</outline>
- add <img> ability to set size over lineHeight
- add <img> align tag like align=top|center|bottom, if not setting use default:bottom

knoxHuang commented on cocos2d/core/components/CCRichText.js:
```diff
@@ -853,6 +863,30 @@ let RichText = cc.Class({
+            if( sprite ) {
+
+                let LineHeight = this.lineHeight;
+                let LineHeightReal = this.lineHeight * 1.26; //我也不懂为什么为什么node高是实际的x1.26, 这值是看编辑器变化算出来的
```
@caryliu1999 这个你知道吗？

caryliu1999 replied:
> @caryliu1999 这个你知道吗？

这里是要获取节点的contentSize，因为节点的size更新是在下一帧渲染时处理的，然后加了留白空间，这里不要写硬编码，可以使用_forceUpdateRenderData更新。

SantyWang replied:
麻烦 @RazgrizHsu  更新一下

RazgrizHsu replied:
这边是为了计算该行应有的y轴偏移，所以才没有取当前label的ContentSize
而且在这for-loop里取label尺寸有另一个问题，就是取得的尺寸有可能是错的
例如若这个RichText里就只有放一张图片，那取得的尺寸就不是非描边label该有的高度

我后来发现实际高度原来是透过 `textUtils.BASELINE_RATIO`算出来的，
例如：`this._labelHeight = (this._lineCount + textUtils.BASELINE_RATIO) * this.lineHeight;`
所以新的commit里改用这种写法，你们看看是否正确

caryliu1999 replied:
> 这边是为了计算该行应有的y轴偏移，所以才没有取当前label的ContentSize
> 而且在这for-loop里取label尺寸有另一个问题，就是取得的尺寸有可能是错的
> 例如若这个RichText里就只有放一张图片，那取得的尺寸就不是非描边label该有的高度
> 
> 我后来发现实际高度原来是透过 `textUtils.BASELINE_RATIO`算出来的，
> 例如：`this._labelHeight = (this._lineCount + textUtils.BASELINE_RATIO) * this.lineHeight;`
> 所以新的commit里改用这种写法，你们看看是否正确

是的，用这个参数也可以，后面我们去掉这个参数的时候也能及时同步修改这里

RazgrizHsu replied:
了解，谢谢 :>

caryliu1999 commented on cocos2d/core/components/CCRichText.js:
```diff
@@ -853,6 +863,30 @@ let RichText = cc.Class({
+            }
+
+            //adjust y for label with outline
+            let outline = label.getComponent( cc.LabelOutline )
```
加上分号，getCompoennt(cc.LabelOutline) 括号内不需要加空格。

RazgrizHsu replied:
好，新的commit里我尽量修改符合原有的代码风格了

caryliu1999 commented on cocos2d/core/components/CCRichText.js:
```diff
@@ -853,6 +863,30 @@ let RichText = cc.Class({
+
+            //adjust y for label with outline
+            let outline = label.getComponent( cc.LabelOutline )
+            if( outline ) label.y = label.y - outline.width;
```
有描边的情况下，Label文本的中线其实是没有变动的，图文混排是要保证文本与图片的中线位置在同一水平线上，这里做了修改之后的效果麻烦确认下是否正确。

RazgrizHsu replied:
是，label本文的中线是没有变动的，
这个组件在蛮早之前我改过一次，当时的对齐跟现在不太一样，
现在改为了每个label都是anchor=0为基础，实际的y轴再透过for-loop带入行高去进行计算
所以有描边的label总是会比没描边的要高，如图：

![image](https://user-images.githubusercontent.com/3889141/72715839-2dfa1800-3bac-11ea-961a-e4ddcc65d69a.png)
我特地打了”单单”两个字来比对有无描边的高度差异
可以看到左边未改的时候，有描边的的确是高了一点点

所以这边我想应该是正确的 Ps. 我是mac原生模似器直接运行后截图的

RazgrizHsu committed:
fix hard code of get node height, add feature <img offset />

- modify some code to follow team code style
- use textUtils.BASELINE_RATIO to get real node height same as other components
- fix when img align=top the offset of y
- add feature of <img offset=12|12,34 /> single value for y, two value for x,y
